tests/installation/select_patterns_and_packages.pm

```diff
@@ -155,113 +141,127 @@ sub package_action {
 
-            if ($selected && $selected->{needle}->has_tag('on-category')) {
-                movedownelseend;
-                check12qtbug;
-                next;
-            }
-            if ($needs_to_be_selected && !$selected) {
-                wait_screen_change {
-                    send_key ' ';
-                    record_info("select", "");
-                };
-                assert_screen 'current-pattern-selected', 5;
-            }
-            # stick to the default patterns. Check if at least 1 dep. issue was displayed
-            $dep_issue = $self->workaround_dependency_issues || $dep_issue;
+    my $counter = 80;
+    # delete special 'all' and 'default' keys from the check
+    delete $patterns{default};
+    delete $patterns{all};
+    while (1) {
+        die "looping for too long" unless ($counter--);
+        my $needs_to_be_selected;
+        my $ret = check_screen('on-pattern', 1);
+        # this variable will only be updated when the pattern list entry is under the cursor
+        # and the pattern is in the PATTERNS variable and there is a needle for that entry
+        my $current_pattern = 'UNKNOWN_PATTERN';
 
-            if (get_var('PATTERNS', '') =~ /default/ && !(get_var('PATTERNS', '') =~ /$current_pattern/)) {
-                $needs_to_be_selected = $selected;
-                record_info("keep default", "");
-            }
-            if (!$needs_to_be_selected && $selected) {
-                send_key ' ';
-                record_info("unselect", "");
-                assert_screen [qw(current-pattern-unselected current-pattern-autoselected)], 8;
+        if ($ret) {    # unneedled pattern
+            for my $p (keys %patterns) {
+                my $sel = 1;
+                if ($p =~ /^-/) {
+                    # this pattern shall be deselected as indicated by '-' prefix
+                    $sel = 0;
+                    $p =~ s/^-//;
+                }
+                if (match_has_tag("pattern-$p")) {
+                    $needs_to_be_selected = $sel;
+                    $current_pattern      = $p;
+                    # add pattern to the set of detected patterns
+                    $processed_patterns{$p} = undef;
+                    delete $patterns{$p};
+                    record_info($current_pattern, $needs_to_be_selected);
+                }
             }
-            movedownelseend;
-            check12qtbug;
+            last if ((get_var('PATTERNS', '') =~ /default/) && !(scalar keys %patterns));
+        }
+
+        $needs_to_be_selected = 1 if get_var('PATTERNS', '') =~ /all/;
+
+        my $selected = check_screen([qw(current-pattern-selected on-category)], 0);
+        if ($selected && $selected->{needle}->has_tag('on-category')) {
+            move_down;
+            next;
+        }
+        if ($needs_to_be_selected && !$selected) {
+            switch_selection('select');
+            assert_screen 'current-pattern-selected', 5;
```

Comment:
Why not to do assertion in the `switch_selection`? It would work pretty well. Also, I would change that method a bit, so we could define action e.g. `switch_selection( select => 1)`  or `switch_selection( action => select )`

Reply:
I passed `action` and `needles` parameters.